src/sentry/api/endpoints/grouping_level_details.py

```diff
@@ -0,0 +1,194 @@
+    for other_group_id, state in GroupHash.objects.filter(
+        project_id=group.project_id, hash__in=levels_overview.parent_hashes
+    ).values_list("group_id", "state"):
+        if other_group_id is not None and state != GroupHash.State.SPLIT:
+            logger.error("grouping_levels.stacked_groups", extra={"group_id": group.id})
+
+    # Mark one hierarchical hash as SPLIT. Note this also prevents it from
+    # being deleted in group deletion.
+    #
+    # We're upserting the hash here to make sure it exists. We have
+    # observed that the materialized hash in postgres is mysteriously lost,
+    # presumably because of secondary grouping or merge/unmerge.
+    grouphash, _created = GroupHash.objects.get_or_create(
+        project_id=group.project_id, hash=levels_overview.current_hash
+    )
+    grouphash.state = GroupHash.State.SPLIT
+    grouphash.group_id = None
+    grouphash.save()
+
+    replacement = HierarchicalUnmergeReplacement(
+        primary_hash=levels_overview.only_primary_hash,
+        filter_hierarchical_hash=levels_overview.current_hash,
+        filter_level=levels_overview.current_level,
+        new_level=id,
+        assume_source_emptied=True,
+        reset_hashes=[],
+    )
+
+    unmerge.delay(
+        project_id=group.project_id,
+        source_id=group.id,
+        destination_id=None,
+        fingerprints=None,
+        actor_id=request.user.id if request.user else None,
+        replacement=replacement,
+    )
+
+
+def _decrease_level(group: Group, id: int, levels_overview: LevelsOverview, request):
+    assert levels_overview.parent_hashes
+    assert id < len(levels_overview.parent_hashes)
+
+    new_materialized_hash = levels_overview.parent_hashes[id]
+
+    # Create new group with parent hash that we want to group by going forward.
+    destination = Group.objects.create(
+        project_id=group.project_id, short_id=group.project.next_short_id()
+    )
+
+    GroupHash.objects.create_or_update(
+        project_id=group.project_id,
+        hash=new_materialized_hash,
+        defaults={"group_id": destination.id},
+    )
+
+    now = datetime.datetime.now()
+
+    # When decreasing the level, events from this + other groups get merged
+    # into one new parent group. This is why we cannot restrict the query by
+    # timestamp here. The query logic is analoguous to
+    # GroupingLevelNewIssuesEndpoint.
+
+    query = (
+        Query("events", Entity("events"))
+        .set_select(
+            [
+                Column("group_id"),
+                Function(
+                    "argMax",
+                    [
+                        Function(
+                            "arraySlice",
+                            [
+                                Column("hierarchical_hashes"),
+                                id + 1,
+                            ],
+                        ),
+                        Column("timestamp"),
```

Comment:
Could it happen that `hierarchical_hashes` is shorter for some events (i.e. the call stack is less deep) than for others, and we then miss some hashes that should be reset?

Reply:
good point, I rewrote this piece of code such that all hashes are returned here for sure